sqlmesh/core/snapshot/evaluator.py

```diff
@@ -953,7 +963,12 @@ def _demote_snapshot(
         environment_naming_info: EnvironmentNamingInfo,
         on_complete: t.Optional[t.Callable[[SnapshotInfoLike], None]],
     ) -> None:
-        adapter = self.adapter
+        adapter = (
+            self._get_adapter(snapshot.model_gateway)
+            if environment_naming_info.gateway_managed_virtual_layer
+            and isinstance(snapshot, Snapshot)
```

Comment:
I don't think I like this approach. Let's not implicitly change the behavior depending on which type was passed, it never ends well. I'd rather extend `SnapshotTableInfo` and `SnapshotInfoLike` interface to support this flag.